smpmgr/common.py

```diff
@@ -39,7 +39,7 @@ def get_custom_smpclient(options: Options, smp_client_cls: Type[TSMPClient]) ->
         logger.info(
             f"Initializing SMPClient with the SMPSerialTransport, {options.transport.port=}"
         )
-        return smp_client_cls(SMPSerialTransport(mtu=options.mtu), options.transport.port)
+        return smp_client_cls(SMPSerialTransport(), options.transport.port)
```

Comment:
I think that this needs to be left in.  Though, it brings up a bigger issue which is that the Serial Transport has been updated to be more 1:1 with the SMP server:

https://github.com/intercreate/smpclient/blob/0185b53a6ac35f662b2aad60c2e90e9fd981bc4d/smpclient/transport/serial.py#L71-L73

Realistically, this needs to be addressed with another feature I've been putting off, which is "per transport" options with reasonable defaults.  I might have to implement that separately and then we rebase this feature branch and see where we're at.

Reply:
Here's the interface I had in my WIP branch:

```
│ --transport-options         TEXT                            Transport options as a JSON      │
│                                                             string, e.g. '{"baudrate":       │
│                                                             115200}'.  Use `smpmgr transport │
│                                                             --help` to see available         │
│                                                             options.                         │
```

```
smpmgr transport                                                              󰯸 smpmgr
Usage: smpmgr transport [OPTIONS] TRANSPORT:{serial|usb}
Try 'smpmgr transport --help' for help.
╭─ Error ──────────────────────────────────────────────────────────────────────────────────────╮
│ Missing argument 'TRANSPORT:{serial|usb}'. Choose from:                                      │
│         serial,                                                                              │
│         usb                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────╯
```

```
smpmgr transport usb                                                          󰯸 smpmgr
Options for --transport-options when using --port
{
  "max_size": 256,
  "line_length": 128,
  "line_buffers": 2
}
```

```
smpmgr transport serial                                                       󰯸 smpmgr
Options for --transport-options when using --port
{
  "max_size": 256,
  "line_length": 128,
  "line_buffers": 2,
  "baudrate": 115200,
  "bytesize": 8,
  "parity": "N",
  "stopbits": 1,
  "timeout": null,
  "xonxoff": false,
  "rtscts": false,
  "write_timeout": null,
  "dsrdtr": false,
  "inter_byte_timeout": null,
  "exclusive": null
}
```

And then usage would be like
```
smpmgr --transport-options '{"max_size": 127, "line_length": 127, "line_buffers": 1}'
```

TBH, it feels like the different transports should be subcommands.  Like
```
smpmgr usb
smpmgr ble
smpmgr udp
smpmgr serial
```
etc.

My JSON idea I think was solely based on convenience for me, not for a user 😢.

Reply:
I think it's safe to update this like so for now:

```python
#if mtu option is supplied:
return smp_client_cls(SMPSerialTransport(max_smp_encoded_frame_size=options.mtu, line_length=options.mtu, line_buffers=1), options.transport.port)
#else
return smp_client_cls(SMPSerialTransport(), options.transport.port)
```

Reply:
Thinking of updating the terminology for users:

- max_smp_encoded_frame_size = smp_size (AKA netbuf size)
- line_length = buffer_size (AKA netbuf size / netbuf count)
- line_buffers = buffer_count (netbuf count)

This terminology should overlap better with BLE transports where:
- smp size = netbuf size
- buffer_size = "MTU - overhead" AKA "max write without response size"
- buffer_count = netbuf count

Because line_length and line_buffers don't make sense for BLE.  Or serial, but I didn't write the spec.

Reply:
> I think it's safe to update like this for now:
> 
> ```python
> #if mtu option is supplied:
> return smp_client_cls(
>     SMPSerialTransport(
>         max_smp_encoded_frame_size=options.mtu,
>         line_length=options.mtu,
>         line_buffers=1
>     ),
>     options.transport.port
> )
> #else
> return smp_client_cls(SMPSerialTransport(), options.transport.port)
> ```

@sgfeniex Are you able to test this with your device?  We don't need to include any breaking changes for now and I think that passing the "MTU" this way will work.

Reply:
Nice speeds! 🔥